src/protocols/multi_party_ecdsa/gg_2020/orchestrate.rs

```diff
@@ -0,0 +1,1241 @@
+    let result = Keys::verify_dlog_proofs(&input.params_s, &input.dlog_proof_vec_s, &input.y_vec_s);
+    if result.is_err() {
+        let err_obj = result.unwrap_err();
+        println!("KeyGen phase 3 checks failed. {:?}", err_obj.clone());
+        Err(err_obj)
+    } else {
+        Ok(())
+    }
+}
+
+macro_rules! write_input {
+    ($index: expr, $stage: expr, $op: expr, $json: expr) => {{
+        if var_os("WRITE_FILE").is_some() {
+            use std::fs::OpenOptions;
+            let mut json_file = OpenOptions::new()
+                .append(true)
+                .open(&format!("{}.txt", $op))
+                .unwrap();
+            let index = $index;
+            let stage = $stage;
+            let op = $op;
+            let json = $json;
+            json_file
+                .write_all(format!("Input {} stage {} index {}\n", op, stage, index).as_bytes())
+                .unwrap();
+            json_file
+                .write_all(format!("{}\n", json).as_bytes())
+                .unwrap();
+        }
+    }};
+}
+macro_rules! write_output {
+    ($index: expr, $stage: expr, $op: expr, $json: expr) => {{
+        if var_os("WRITE_FILE").is_some() {
+            use std::fs::OpenOptions;
+            let mut json_file = OpenOptions::new()
+                .append(true)
+                .open(&format!("{}.txt", $op))
+                .unwrap();
+            let index = $index;
+            let stage = $stage;
+            let op = $op;
+            let json = $json;
+            json_file
+                .write_all(format!("Output {} stage {} index {}\n", op, stage, index).as_bytes())
+                .unwrap();
+            json_file
+                .write_all(format!("{}\n", json).as_bytes())
+                .unwrap();
+        }
+    }};
+}
+
+// The Distributed key generation protocol can work with a broadcast channel.
+// All the messages are exchanged p2p.
+// On the contrary, the key generation process can be orchestrated as below.
+// All the participants do some work on each stage and return some data.
+// This data needs to be filtered/collated and sent back as an input to the next stage.
+// This test helper is just a demonstration of the same.
+//
+pub fn keygen_orchestrator(params: Parameters) -> Result<KeyPairResult, ErrorType> {
+    let op = "keygen".to_string();
+    if var_os("WRITE_FILE").is_some() {
+        File::create(&format!("{}.txt", &op)).unwrap();
+    }
+    //
+    // Values to be kept private(Each value needs to be encrypted with a key only known to that
+    // party):
+    // 1. party_keys.u_i
+    // 2. party_keys.dk
+    let mut party_keys_vec_l = vec![];
+    // Nothing private in the commitment values.
+    let mut bc1_vec_l = vec![];
+    // Values to be kept private(Each value needs to be encrypted with a key only known to that
+    // party):
+    // 1. decommitment values in KeyGenDecommitMessage1 need to be encrypted until they are sent
```

Comment:
why decommitments need to be encrypted: 
- each party generates its own decommitment and keeps it private until the relevant round
- at the relevant round the party broadcast the decom and other parties verify

Reply:
If the party encrypts it and throws it out that makes party stateless. All it needs to keep is the encryption key. This stage api can then serve multiple key generations in parallel in case you call it in different threads in parallel